### PR TITLE
Evaluation tool

### PR DESCRIPTION
**Pull request by hgaiser**
This PR merges previous work from https://github.com/fizyr/keras-retinanet/pull/196 and previous work done on the `evaluation` branch. This tool should provide `mAP` evaluation scores for arbitrary generators (and therefore datasets).

The metric computed is largely derived from [`py-faster-rcnn`](https://github.com/rbgirshick/py-faster-rcnn/blob/master/lib/datasets/voc_eval.py). Compared to the COCO evaluation it should theoretically the same as AP@0.5. Testing on the pretrained COCO model gives an `mAP` of 0.507, whereas the COCO evaluation gives an `AP@0.5` of 0.513. I'm not sure where this discrepancy comes from, this is something that could be further investigated in the future.

@awilliamson could you have a look too ?

### CODE REVIEW

File: keras_retinanet/utils/eval.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        # find the order with which to sort the scores
+        scores_sort = np.argsort(-scores)[:max_detections]
+
+        # select detections
+        image_boxes      = detections[0, indices[0][scores_sort], :4]
+        image_scores     = np.expand_dims(detections[0, indices[0][scores_sort], 4 + indices[1][scores_sort]], axis=1)
+        image_detections = np.append(image_boxes, image_scores, axis=1)
+        image_predicted_labels = indices[1][scores_sort]
+
+        if save_path:
```
Comment (de-vri-es) [inline keras_retinanet/utils/eval.py]: I would go for `if save_path is not None:` here. Otherwise, `--save-path ''` on the command line doesn't save any images. I would expect the images to be saved to the current working directory. I would be a bit annoyed if it turns out afterwards that no images were saved.

`--save-path .` would be a workaround, but I don't see a reason to special case empty strings here.

**Comment by de-vri-es** [comment]
Code looks good in general. I didn't test it manually, I do think we should do that before merging (if you haven't already). I do have a few remarks (besides the in-line comment):

* There is no documentation for the added functions. Even for internal functions, some documentation on the intended behaviour and structure of data would be very useful for future maintenance.

* There are quite some command line options in common with the train script. I'm not sure that we should share these between the scripts, but if not we should manually make sure they stay in sync with each other in the future.

* It would also be nice to be able to specify a random transform generator on the command line to see the performance of the network on augmented data (and to test the random transform generator ^_^). We can add this in a separate PR to both training and testing scripts though.

* There doesn't seem to be an evaluation callback for use with training as discussed in #225. It can be a separate PR, but if nobody is working on it right now I'd like to re-open the issue so that we don't forget it.

File: keras_retinanet/utils/eval.py (hunk 1 above)
Comment (hgaiser) [reply keras_retinanet/utils/eval.py]: Agreed, edited.

**Comment by hgaiser** [comment]
Thanks for the review. Fair point on the documentation, I will add it soon. Also thanks for reminding me of the callback, should be fairly straightforward to add that. I'll let you know when I'm done updating.

**Comment by hgaiser** [comment]
Alright I updated it with documentation and a callback. Can you review @de-vri-es ?

File: keras_retinanet/bin/train.py [hunk 2]
```diff
@@ -208,7 +210,8 @@ def parse_args(args):
     parser.add_argument('--steps',         help='Number of steps per epoch.', type=int, default=10000)
     parser.add_argument('--snapshot-path', help='Path to store snapshots of models during training (defaults to \'./snapshots\')', default='./snapshots')
     parser.add_argument('--no-snapshots',  help='Disable saving snapshots.', dest='snapshots', action='store_false')
-    parser.set_defaults(snapshots=True)
+    parser.add_argument('--no-evaluation', help='Disable per epoch evaluation.', dest='evaluation', action='store_false')
+    parser.set_defaults(snapshots=True, evaluation=True)
```
Comment (de-vri-es) [inline keras_retinanet/bin/train.py]: Same as above: the `set_defaults` is not needed.

File: keras_retinanet/bin/evaluate.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    csv_parser.add_argument('classes', help='Path to a CSV file containing class label mapping.')
+
+    parser.add_argument('model',             help='Path to RetinaNet model.')
+    parser.add_argument('--gpu',             help='Id of the GPU to use (as reported by nvidia-smi).')
+    parser.add_argument('--score-threshold', help='Threshold on score to filter detections with (defaults to 0.05).', default=0.05, type=float)
+    parser.add_argument('--iou-threshold',   help='IoU Threshold to count for a positive detection (defaults to 0.5).', default=0.5, type=float)
+    parser.add_argument('--max-detections',  help='Max Detections per image (defaults to 100).', default=100, type=int)
+    parser.add_argument('--save-path',       help='Path for saving images with detections.')
+
+    parser.set_defaults(snapshots=True)
```
Comment (de-vri-es) [inline keras_retinanet/bin/evaluate.py]: The `set_defaults` is not needed. `store_true` and `store_false` already set the default value to the opposite value: https://docs.python.org/3/library/argparse.html#action

**Comment by hgaiser** [comment]
Ah I see, I didn't know that. I updated it now, shall I merge it then?

**Comment by de-vri-es** [comment]
Yeah, good with me =]

**Comment by Weixing-Zhang** [comment]
@de-vri-es 
Thanks a lot! Please allow me to have you in my paper's acknowledgements as same as Hans Gaiser. Maarten de Vries? Let me know if I misspell your name.

**Comment by de-vri-es** [comment]
That is correct. Thanks :)
